extensions/amp-analytics/0.1/transport.js

```diff
@@ -118,13 +144,272 @@ export class Transport {
+   * this particular creative to the cross-domain iframe
+   * @returns {string}
+   * @private
+   */
+  static createUniqueId_() {
+    let newId;
+    do {
+      newId = String(Math.random()).substr(2);
+    } while (Transport.usedIds_[newId]);
+    Transport.usedIds_[newId] = true;
+    return newId;
+  }
+
+  /**
+   * Indicate that a cross-domain frame is ready to receive messages, and
+   * send all messages that were previously queued for it.
+   * @param {!string} frameUrl The URL of the frame
+   * @private
+   */
+  static setIsReady_(frameUrl) {
+    const frameData = Transport.crossDomainIframes_[frameUrl];
+    frameData.isReady = true;
+    Transport.sendQueuedMessagesToCrossDomainIframe_(frameData);
+  }
+
+  /**
+   * Sends an Amp Analytics trigger event to a vendor's cross-domain iframe,
+   * or queues the message if the frame is not yet ready to receive messages.
+   * @param {string} request
+   * @param {!Object<string, string>} transportOptions
+   * @private
+   */
+  sendRequestUsingCrossDomainIframe_(request, transportOptions) {
+    const frameData = Transport.crossDomainIframes_[transportOptions['iframe']];
+    dev().assert(frameData, 'Trying to send message to non-existent frame');
+    this.enqueueMessageForCrossDomainIframe_(
+      /** @type {!Object<string,*>} */ (frameData),
+      MessageTypes.ampAnalytics3pEvent, request);
+  }
+
+  /**
+   * Sends extra data (from the amp-analytics config object) to the
+   * cross-domain iframe.
+   * @param {!string} frameUrl The URL of the frame to send the data to
+   * @param {string=} opt_extraData The data to send to the frame
+   */
+  informCrossDomainIframeOfNewCreative_(frameUrl, opt_extraData) {
+    opt_extraData = opt_extraData || ''; // Still send the message to
+    // indicate there is a new creative
+    const frameData = Transport.crossDomainIframes_[frameUrl];
+    dev().assert(frameData, 'Trying to send message to non-existent frame');
+    this.enqueueMessageForCrossDomainIframe_(
+      /** @type {!Object<string,*>} */ (frameData),
+      MessageTypes.ampAnalytics3pNewCreative, opt_extraData);
+  }
+
+  /**
+   * Enqueues a message (event or extra data) to be sent to a cross-domain
+   * iframe.
+   * @param {!Object<string,*>} frameData  The cross-domain iframe
+   * @param {!string} messageType The type of the message (see MessageTypes)
+   * @param {!string} message
+   * @private
+   */
+  enqueueMessageForCrossDomainIframe_(frameData, messageType, message) {
+    if (frameData.messageQueue.length > MAX_QUEUE_SIZE_) {
+      dev().warn(TAG_, 'Queue has exceeded maximum size');
+    }
+    const messageObject = {
+      senderId: this.id_,
+      type: messageType,
+    };
+    messageObject[messageType] = message;
+    frameData.messageQueue.push(messageObject);
+    if (frameData.sendTimer) {
+      return; // Timer is already about to fire, no need to set a new one
+    }
+    frameData.sendTimer = Transport.timer_.delay(() => {
```

Comment:
So you mean, send the first one immediately, and then set an invariant "I will not send again for at least 100ms", and if subsequent ones come in during that 100ms queue them until that time has elapsed? I like this.
(Currently implementation destroys the timer about 4 lines down from here, upon it firing.)